File: clients/go/transform.go
package sharptown

import (
	"context"
	"fmt"
	"strconv"
	"strings"
)

// Transform is a chainable builder for a single image transformation. Operation methods
// return the same *Transform, so they compose fluently; a terminal (Do, Bytes, Save) runs
// the request. The first invalid operation is captured and returned by the terminal.
type Transform struct {
	client *Client
	input  Input
	ops    map[string]any
	err    error
}

// Operations returns the canonical operation set accumulated so far.
func (t *Transform) Operations() map[string]any { return t.ops }

// Err returns the first validation error captured, if any.
func (t *Transform) Err() error { return t.err }

// Resize sets the target width and height.
func (t *Transform) Resize(width, height int) *Transform {
	return t.putUint("width", width, "width").putUint("height", height, "height")
}

// Width sets the target width only.
func (t *Transform) Width(value int) *Transform { return t.putUint("width", value, "width") }

// Height sets the target height only.
func (t *Transform) Height(value int) *Transform { return t.putUint("height", value, "height") }

// Crop crops a rectangle (left, top, width, height).
func (t *Transform) Crop(left, top, width, height int) *Transform {
	if t.err != nil {
		return t
	}
	for name, value := range map[string]int{"crop.left": left, "crop.top": top, "crop.width": width, "crop.height": height} {
		if value < 0 {
			return t.fail(fmt.Errorf("invalid %s: expected a non-negative integer, got %d", name, value))
		}
	}
	t.ops["crop"] = fmt.Sprintf("%d,%d,%d,%d", left, top, width, height)
	return t
}

// SmartCrop crops to the salient region when resizing.
func (t *Transform) SmartCrop() *Transform { return t.set("smartCrop", true) }

// Fit sets the resize fit mode: cover, contain, fill, inside, outside.
func (t *Transform) Fit(mode string) *Transform {
	m := strings.ToLower(mode)
	if !contains(FitModes, m) {
		return t.fail(fmt.Errorf("unsupported fit %q. Supported: %s", mode, strings.Join(FitModes, ", ")))
	}
	return t.set("fit", m)
}

// Background sets the background colour for fit=contain.
func (t *Transform) Background(color string) *Transform { return t.set("background", color) }

// DPR sets the device pixel ratio (multiplies the target size).
func (t *Transform) DPR(value float64) *Transform { return t.putRange("dpr", value, "dpr", 0.1, 5) }

// AspectRatio sets the target aspect ratio (width / height).
func (t *Transform) AspectRatio(value float64) *Transform {
	return t.putRange("aspectRatio", value, "aspectRatio", 0.0001, 1000)
}

// AutoOrient rotates by EXIF orientation.
func (t *Transform) AutoOrient() *Transform { return t.set("autoOrient", true) }

// Rotate rotates by the given degrees.
func (t *Transform) Rotate(degrees int) *Transform { return t.set("rotate", degrees) }

// Flip flips horizontally.
func (t *Transform) Flip() *Transform { return t.set("flip", true) }

// Blur applies a Gaussian blur of the given sigma/radius.
func (t *Transform) Blur(sigma int) *Transform { return t.putUint("blur", sigma, "blur") }

// Tint tints the image with the given RGB channels (each 0–255).
func (t *Transform) Tint(r, g, b int) *Transform {
	return t.putColor("r", r, "r").putColor("g", g, "g").putColor("b", b, "b")
}

// Grayscale desaturates the image.
func (t *Transform) Grayscale() *Transform { return t.set("grayscale", true) }

// Greyscale is a British alias of Grayscale.
func (t *Transform) Greyscale() *Transform { return t.Grayscale() }

// RemoveAlpha removes the alpha channel.
func (t *Transform) RemoveAlpha() *Transform { return t.set("removeAlpha", true) }

// EnsureAlpha ensures an alpha channel exists.
func (t *Transform) EnsureAlpha() *Transform { return t.set("ensureAlpha", true) }

// Brightness adjusts brightness, -100–100.
func (t *Transform) Brightness(value float64) *Transform {
	return t.putRange("brightness", value, "brightness", -100, 100)
}

// Contrast adjusts contrast, -100–100.
func (t *Transform) Contrast(value float64) *Transform {
	return t.putRange("contrast", value, "contrast", -100, 100)
}

// Saturation adjusts saturation, 0–2.
func (t *Transform) Saturation(value float64) *Transform {
	return t.putRange("saturation", value, "saturation", 0, 2)
}

// Exposure adjusts exposure in EV stops, -3–3.
func (t *Transform) Exposure(value float64) *Transform {
	return t.putRange("exposure", value, "exposure", -3, 3)
}

// Hue rotates hue in degrees, 0–360.
func (t *Transform) Hue(value float64) *Transform { return t.putRange("hue", value, "hue", 0, 360) }

// Gamma applies gamma correction, 1.0–3.0.
func (t *Transform) Gamma(value float64) *Transform { return t.putRange("gamma", value, "gamma", 1, 3) }

// Colorize maps the image to shades of one colour.
func (t *Transform) Colorize(color string) *Transform { return t.set("colorize", color) }

// Sepia applies a sepia tone, intensity 0–1.
func (t *Transform) Sepia(intensity float64) *Transform {
	return t.putRange("sepia", intensity, "sepia", 0, 1)
}

// Invert inverts colours.
func (t *Transform) Invert() *Transform { return t.set("invert", true) }

// Threshold binarises the image at the given threshold, 0–255.
func (t *Transform) Threshold(value int) *Transform {
	return t.putIntRange("threshold", value, "threshold", 0, 255)
}

// Sharpen sharpens the image with the given sigma, 0–5.
func (t *Transform) Sharpen(sigma float64) *Transform {
	return t.putRange("sharpen", sigma, "sharpen", 0, 5)
}

// OilPaint applies an oil-paint effect with the given window size, 1–25.
func (t *Transform) OilPaint(size int) *Transform {
	return t.putIntRange("oilPaint", size, "oilPaint", 1, 25)
}

// Quality sets the output quality 1–100 (applies with Convert).
func (t *Transform) Quality(value int) *Transform {
	return t.putIntRange("quality", value, "quality", 1, 100)
}

// Progressive enables progressive (interlaced) output.
func (t *Transform) Progressive() *Transform { return t.set("progressive", true) }

// StripMetadata strips EXIF/metadata (the server default).
func (t *Transform) StripMetadata() *Transform { return t.set("stripMetadata", true) }

// KeepMetadata keeps EXIF/metadata in the output.
func (t *Transform) KeepMetadata() *Transform { return t.set("stripMetadata", false) }

// Convert sets the output format (webp, png, jpg, jpeg, avif, gif, heif).
func (t *Transform) Convert(format string) *Transform {
	f := strings.ToLower(format)
	if !contains(SupportedFormats, f) {
		return t.fail(fmt.Errorf("unsupported format %q. Supported: %s", format, strings.Join(SupportedFormats, ", ")))
	}
	return t.set("convertTo", f)
}

// ToFormat is an alias of Convert.
func (t *Transform) ToFormat(format string) *Transform { return t.Convert(format) }

// Do runs the request and returns the full Response.
func (t *Transform) Do(ctx context.Context) (*Response, error) {
	if t.err != nil {
		return nil, t.err
	}
	req := &Request{
		BaseURL:    t.client.baseURL,
		Headers:    t.client.headers,
		Input:      t.input,
		Operations: t.ops,
		Timeout:    t.client.timeout,
		HTTPClient: t.client.httpClient,
	}
	if t.client.timeout > 0 {
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, t.client.timeout)
		defer cancel()
	}
	return t.client.transport.Transform(ctx, req)
}

// Bytes runs the request and returns the raw image bytes.
func (t *Transform) Bytes(ctx context.Context) ([]byte, error) {
	resp, err := t.Do(ctx)
	if err != nil {
		return nil, err
	}
	return resp.Body, nil
}

// Save runs the request and writes the result to a file.
func (t *Transform) Save(ctx context.Context, path string) error {
	resp, err := t.Do(ctx)
	if err != nil {
		return err
	}
	return resp.Save(path)
}

func (t *Transform) set(key string, value any) *Transform {
	if t.err == nil {
		t.ops[key] = value
	}
	return t
}

func (t *Transform) fail(err error) *Transform {
	if t.err == nil {
		t.err = &Error{Message: err.Error()}
	}
	return t
}

func (t *Transform) putUint(key string, value int, field string) *Transform {
	if t.err != nil {
		return t
	}
	if value < 0 {
		return t.fail(fmt.Errorf("invalid %s: expected a non-negative integer, got %d", field, value))
	}
	return t.set(key, value)
}

func (t *Transform) putColor(key string, value int, field string) *Transform {
	if t.err != nil {
		return t
	}
	if value < 0 || value > 255 {
		return t.fail(fmt.Errorf("invalid %s: expected 0-255, got %d", field, value))
	}
	return t.set(key, value)
}

func (t *Transform) putIntRange(key string, value int, field string, min, max int) *Transform {
	if t.err != nil {
		return t
	}
	if value < min || value > max {
		return t.fail(fmt.Errorf("invalid %s: expected %d-%d, got %d", field, min, max, value))
	}
	return t.set(key, value)
}

func (t *Transform) putRange(key string, value float64, field string, min, max float64) *Transform {
	if t.err != nil {
		return t
	}
	if value < min || value > max {
		return t.fail(fmt.Errorf("invalid %s: expected %s-%s, got %s", field, ftoa(min), ftoa(max), ftoa(value)))
	}
	return t.set(key, value)
}

func ftoa(value float64) string { return strconv.FormatFloat(value, 'f', -1, 64) }

File: clients/flutter/lib/src/transform.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:http/http.dart' as http;

import 'input.dart';
import 'operations.dart';
import 'response.dart';
import 'transport/transport.dart';
import 'watermark.dart';

/// A chainable builder for transforming a single image. Every operation method returns
/// `this`, so calls compose fluently; a terminal ([bytes], [response], [save]) runs the
/// request.
///
/// Operations are validated as you build — an out-of-range value or unsupported format
/// throws a `SharptownError` immediately, before any request is sent.
///
/// ```dart
/// final data = await st
///     .transform(ImageInput.file(file))
///     .resize(800, 600)
///     .blur(3)
///     .grayscale()
///     .convert('webp')
///     .bytes();
/// ```
class TransformBuilder {
  /// Creates a builder. Prefer `SharptownClient.transform` over calling this directly.
  TransformBuilder({
    required Transport transport,
    required String baseUrl,
    required Map<String, String> headers,
    required Duration timeout,
    required http.Client httpClient,
    required ImageInput input,
  })  : _transport = transport,
        _baseUrl = baseUrl,
        _headers = headers,
        _timeout = timeout,
        _httpClient = httpClient,
        _input = input;

  final Transport _transport;
  final String _baseUrl;
  final Map<String, String> _headers;
  final Duration _timeout;
  final http.Client _httpClient;
  final ImageInput _input;
  final Map<String, Object?> _ops = {};
  final List<CompositeMark> _marks = [];

  /// The canonical operation set accumulated so far.
  Map<String, Object?> get operations => Map.unmodifiable(_ops);

  /// Trims uniform edges. Pass no argument for the default threshold, or a 1–255 threshold.
  TransformBuilder trim([int? threshold]) =>
      _set('trim', threshold == null ? true : Operations.positiveInt(threshold, 'trim'));

  /// Makes a colour transparent (chroma key). Pass the colour (`#rrggbb`, `r,g,b` or a name)
  /// and an optional tolerance percentage (0–100, default 12). Applied on the REST server.
  TransformBuilder chromaKey(String color, [int? tolerance]) =>
      _set('chromaKey', tolerance == null ? color : '$color;$tolerance');

  /// Overlays a [Watermark] (image) or [Textmark] (text) onto the result. Call it more than
  /// once to stack overlays; they are composited in order. Applied on the REST server.
  TransformBuilder composite(CompositeMark mark) {
    _marks.add(mark);
    return this;
  }

  /// Sets the target [width] and/or [height].
  TransformBuilder resize([int? width, int? height]) {
    if (width != null) {
      _ops['width'] = Operations.positiveInt(width, 'width');
    }
    if (height != null) {
      _ops['height'] = Operations.positiveInt(height, 'height');
    }
    return this;
  }

  /// Sets the target width only.
  TransformBuilder width(int value) =>
      _set('width', Operations.positiveInt(value, 'width'));

  /// Sets the target height only.
  TransformBuilder height(int value) =>
      _set('height', Operations.positiveInt(value, 'height'));

  /// Crops a rectangle ([left], [top], [width], [height]).
  TransformBuilder crop(int left, int top, int width, int height) {
    Operations.positiveInt(left, 'crop.left');
    Operations.positiveInt(top, 'crop.top');
    Operations.positiveInt(width, 'crop.width');
    Operations.positiveInt(height, 'crop.height');
    return _set('crop', '$left,$top,$width,$height');
  }

  /// Crops to the salient region when resizing.
  TransformBuilder smartCrop() => _set('smartCrop', true);

  /// Sets the resize fit mode: `cover`, `contain`, `fill`, `inside`, `outside`.
  TransformBuilder fit(String mode) =>
      _set('fit', Operations.assertFit(mode.toLowerCase()));

  /// Sets the background colour used by `fit: contain`.
  TransformBuilder background(String color) => _set('background', color);

  /// Sets the device pixel ratio (multiplies the target size), 0.1–5.
  TransformBuilder dpr(num value) =>
      _set('dpr', Operations.range(value, 'dpr', 0.1, 5));

  /// Sets the target aspect ratio (width / height).
  TransformBuilder aspectRatio(num value) =>
      _set('aspectRatio', Operations.range(value, 'aspectRatio', 0.0001, 1000));

  /// Rotates by EXIF orientation.
  TransformBuilder autoOrient() => _set('autoOrient', true);

  /// Rotates by the given [degrees].
  TransformBuilder rotate(int degrees) => _set('rotate', degrees);

  /// Flips horizontally.
  TransformBuilder flip() => _set('flip', true);

  /// Adjusts brightness, -100–100.
  TransformBuilder brightness(num value) =>
      _set('brightness', Operations.range(value, 'brightness', -100, 100));

  /// Adjusts contrast, -100–100.
  TransformBuilder contrast(num value) =>
      _set('contrast', Operations.range(value, 'contrast', -100, 100));

  /// Adjusts saturation, 0–2.
  TransformBuilder saturation(num value) =>
      _set('saturation', Operations.range(value, 'saturation', 0, 2));

  /// Adjusts exposure in EV stops, -3–3.
  TransformBuilder exposure(num value) =>
      _set('exposure', Operations.range(value, 'exposure', -3, 3));

  /// Rotates hue in degrees, 0–360.
  TransformBuilder hue(num value) =>
      _set('hue', Operations.range(value, 'hue', 0, 360));

  /// Applies gamma correction, 1.0–3.0.
  TransformBuilder gamma(num value) =>
      _set('gamma', Operations.range(value, 'gamma', 1, 3));

  /// Maps the image to shades of one [color].
  TransformBuilder colorize(String color) => _set('colorize', color);

  /// Tints the image with the given RGB channels (each 0–255).
  TransformBuilder tint(int r, int g, int b) {
    _ops['r'] = Operations.color(r, 'r');
    _ops['g'] = Operations.color(g, 'g');
    _ops['b'] = Operations.color(b, 'b');
    return this;
  }

  /// Desaturates the image.
  TransformBuilder grayscale() => _set('grayscale', true);

  /// British alias of [grayscale].
  TransformBuilder greyscale() => grayscale();

  /// Applies a Gaussian blur of the given [sigma]/radius.
  TransformBuilder blur(int sigma) =>
      _set('blur', Operations.positiveInt(sigma, 'blur'));

  /// Sharpens the image with the given [sigma], 0–5.
  TransformBuilder sharpen(num sigma) =>
      _set('sharpen', Operations.range(sigma, 'sharpen', 0, 5));

  /// Applies a sepia tone, [intensity] 0–1.
  TransformBuilder sepia(num intensity) =>
      _set('sepia', Operations.range(intensity, 'sepia', 0, 1));

  /// Inverts colours.
  TransformBuilder invert() => _set('invert', true);

  /// Binarises the image at the given [value], 0–255.
  TransformBuilder threshold(int value) =>
      _set('threshold', Operations.intRange(value, 'threshold', 0, 255));

  /// Applies an oil-paint effect with the given window [size], 1–25.
  TransformBuilder oilPaint(int size) =>
      _set('oilPaint', Operations.intRange(size, 'oilPaint', 1, 25));

  /// Removes the alpha channel.
  TransformBuilder removeAlpha() => _set('removeAlpha', true);

  /// Ensures an alpha channel exists.
  TransformBuilder ensureAlpha() => _set('ensureAlpha', true);

  /// Sets the output [format] (webp, png, jpg, jpeg, avif, gif, heif).
  TransformBuilder convert(String format) =>
      _set('convertTo', Operations.assertFormat(format.toLowerCase()));

  /// Alias of [convert].
  TransformBuilder toFormat(String format) => convert(format);

  /// Sets the output quality 1–100 (applies with [convert]).
  TransformBuilder quality(int value) =>
      _set('quality', Operations.intRange(value, 'quality', 1, 100));

  /// Enables progressive (interlaced) output.
  TransformBuilder progressive() => _set('progressive', true);

  /// Strips EXIF/metadata (the server default).
  TransformBuilder stripMetadata() => _set('stripMetadata', true);

  /// Keeps EXIF/metadata in the output.
  TransformBuilder keepMetadata() => _set('stripMetadata', false);

  /// Runs the request and returns the full [TransformResponse].
  Future<TransformResponse> response() {
    final operations = Map<String, Object?>.of(_ops);
    final attachments = <List<int>>[];
    if (_marks.isNotEmpty) {
      final specs = <Map<String, Object?>>[];
      for (final mark in _marks) {
        final resolved = mark.resolve();
        final spec = resolved.spec;
        if (resolved.bytes != null) {
          spec['ref'] = attachments.length;
          attachments.add(resolved.bytes!);
        }
        specs.add(spec);
      }
      operations['composite'] = jsonEncode(specs);
    }

    return _transport.transform(TransformRequest(
      baseUrl: _baseUrl,
      headers: _headers,
      input: _input,
      operations: operations,
      timeout: _timeout,
      httpClient: _httpClient,
      attachments: attachments,
    ));
  }

  /// Runs the request and returns the raw image bytes.
  Future<Uint8List> bytes() async => (await response()).bytes;

  /// Runs the request and writes the result to [path], returning the written [File].
  Future<File> save(String path) async => (await response()).save(path);

  TransformBuilder _set(String key, Object value) {
    _ops[key] = value;
    return this;
  }
}
